test/rubygems/test_gem_remote_fetcher_s3.rb

```diff
@@ -8,6 +8,96 @@
 class TestGemRemoteFetcherS3 < Gem::TestCase
   include Gem::DefaultUserInteraction
 
+  class FakeGemRequest < Gem::Request
+    attr_reader :last_request, :uri
+
+    # Override perform_request to stub things
+    def perform_request(request)
+      @last_request = request
+      @response
+    end
+
+    def set_response(response)
+      @response = response
+    end
+  end
+
+  class FakeS3URISigner < Gem::S3URISigner
+    # Convenience method to output the recent aws iam queries made in tests
+    # this outputs the verb, path, and any non-generic headers
+    def recent_aws_query_logs
+      sreqs = @aws_iam_calls.map do |c|
+        r = c.last_request
+        s = +"#{r.method} #{c.uri}\n"
+        r.each_header do |key, v|
+          # Only include headers that start with x-
+          next unless key.start_with?("x-")
+          s << "    #{key}=#{v}\n"
+        end
+        s
+      end
+
+      sreqs.join("")
+    end
+
+    def initialize(uri)
+      @aws_iam_calls = []
+      super
+    end
+
+    def ec2_iam_request(uri, verb)
+      fake_s3_request = FakeGemRequest.new(uri, verb, nil, nil)
+      @aws_iam_calls << fake_s3_request
+
+      case uri.to_s
+      when "http://169.254.169.254/latest/api/token"
+        if $imdsv2_token_failure
```

Comment:
What about to set at least class shared variable instead of global one to control this? `FakeS3URISigner.return_token = "mysecrettoken"` => returns token, `FakeS3URISigner.return_token = false` 401 instead and reset before each test + set whenever failure is needed in `with_imds_v2_failure` method?